File: ts/src/provision.ts
/**
 * Provisioning primitives: create, rotate and revoke identities.
 *
 * Building blocks for producers — provisioning scripts, development tooling,
 * operators. They generate keypairs and maintain the trust bundle; consumers
 * only ever read the resulting files via {@link Identity}.
 *
 * A provisioned root directory looks like:
 *
 *     <root>/
 *       bundle.json                # the trust bundle (distribute to everyone)
 *       services/<name>/id         # the service's SPIFFE ID
 *       services/<name>/key.pem    # the service's private key (deliver only to it)
 */
import { chmodSync, existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs"
import { join } from "node:path"

import { Bundle, BUNDLE_VERSION, type BundleDocument } from "./bundle.js"
import { SpiffileError } from "./errors.js"
import {
  DIR_BUNDLE_FILENAME,
  DIR_ID_FILENAME,
  DIR_KEY_FILENAME,
  ENV_BUNDLE_FILE,
  ENV_ID_FILE,
  ENV_KEY_FILE,
  Identity,
} from "./identity.js"
import { generatePrivateKey, privateKeyToPem, publicJwk, type Jwk } from "./keys.js"
import { SpiffeId } from "./spiffeId.js"

const SERVICES_DIRNAME = "services"

/** Create an empty provisioning root with an empty bundle. */
export function initRoot(root: string, trustDomain: string): string {
  const bundlePath = join(root, DIR_BUNDLE_FILENAME)
  if (existsSync(bundlePath)) {
    throw new SpiffileError(`already initialized: ${bundlePath}`)
  }
  mkdirSync(join(root, SERVICES_DIRNAME), { recursive: true })
  writeBundleDocument(bundlePath, {
    trust_domain: trustDomain,
    spiffile_version: BUNDLE_VERSION,
    identities: {},
  })
  return root
}

/**
 * Generate a keypair for a service and register its public key in the
 * bundle. Idempotent: a service that already has a key keeps it.
 */
export function addService(root: string, name: string): SpiffeId {
  const bundle = readBundle(root)
  const spiffeId = SpiffeId.parse(`spiffe://${bundle.trustDomain}/${name}`)

  const serviceDir = join(root, SERVICES_DIRNAME, name)
  const keyPath = join(serviceDir, DIR_KEY_FILENAME)
  if (existsSync(keyPath)) {
    return spiffeId
  }

  mkdirSync(serviceDir, { recursive: true })
  const privateKey = generatePrivateKey()
  writeFileSync(keyPath, privateKeyToPem(privateKey))
  chmodSync(keyPath, 0o600)
  writeFileSync(join(serviceDir, DIR_ID_FILENAME), `${spiffeId}\n`)

  setKeys(root, bundle, spiffeId.toString(), [publicJwk(privateKey)])
  return spiffeId
}

/**
 * Generate a new keypair for a service. With keepOld (the default) the
 * previous public key stays in the bundle so in-flight tokens and stale
 * bundle copies keep working.
 */
export function rotateService(root: string, name: string, keepOld = true): SpiffeId {
  const bundle = readBundle(root)
  const spiffeId = SpiffeId.parse(`spiffe://${bundle.trustDomain}/${name}`)

  const keyPath = join(root, SERVICES_DIRNAME, name, DIR_KEY_FILENAME)
  if (!existsSync(keyPath)) {
    throw new SpiffileError(`unknown service '${name}': ${keyPath} does not exist`)
  }

  const privateKey = generatePrivateKey()
  const newJwk = publicJwk(privateKey)
  const existing = keepOld ? (bundle.identities.get(spiffeId.toString()) ?? []) : []
  const keys = [newJwk, ...existing.filter((jwk) => jwk.kid !== newJwk.kid)]

  writeFileSync(keyPath, privateKeyToPem(privateKey))
  chmodSync(keyPath, 0o600)
  setKeys(root, bundle, spiffeId.toString(), keys)
  return spiffeId
}

/** Remove a service's keys from the bundle (revocation) and its directory. */
export function removeService(root: string, name: string): void {
  const bundle = readBundle(root)
  const doc = bundle.toDocument()
  delete doc.identities[`spiffe://${bundle.trustDomain}/${name}`]
  writeBundleDocument(join(root, DIR_BUNDLE_FILENAME), doc)
  rmSync(join(root, SERVICES_DIRNAME, name), { recursive: true, force: true })
}

/** The environment variables that point a consumer at its identity files. */
export function serviceEnv(root: string, name: string): Record<string, string> {
  const serviceDir = join(root, SERVICES_DIRNAME, name)
  return {
    [ENV_ID_FILE]: join(serviceDir, DIR_ID_FILENAME),
    [ENV_KEY_FILE]: join(serviceDir, DIR_KEY_FILENAME),
    [ENV_BUNDLE_FILE]: join(root, DIR_BUNDLE_FILENAME),
  }
}

/** Load a provisioned service's {@link Identity} directly (tests, tooling). */
export function loadIdentity(root: string, name: string): Identity {
  const env = serviceEnv(root, name)
  return Identity.fromFiles(env[ENV_ID_FILE], env[ENV_KEY_FILE], env[ENV_BUNDLE_FILE])
}

// -- internals ---------------------------------------------------------------

function readBundle(root: string): Bundle {
  return Bundle.fromDocument(JSON.parse(readFileSync(join(root, DIR_BUNDLE_FILENAME), "utf8")))
}

function setKeys(root: string, bundle: Bundle, spiffeId: string, keys: Jwk[]): void {
  const doc = bundle.toDocument()
  doc.identities[spiffeId] = { keys }
  writeBundleDocument(join(root, DIR_BUNDLE_FILENAME), doc)
}

function writeBundleDocument(path: string, doc: BundleDocument): void {
  writeFileSync(path, `${JSON.stringify(doc, null, 2)}\n`)
}

File: go/provision.go
package spiffile

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// Provisioning primitives: create, rotate and revoke identities.
//
// These functions are the building blocks for producers — provisioning
// scripts, development tooling, operators. They generate keypairs and
// maintain the trust bundle; consumers only ever read the resulting files.
//
// A provisioned root directory looks like:
//
//	<root>/
//	  bundle.json                # the trust bundle (distribute to everyone)
//	  services/<name>/id         # the service's SPIFFE ID
//	  services/<name>/key.pem    # the service's private key (deliver only to it)

const (
	servicesDirname = "services"
)

// InitRoot creates an empty provisioning root with an empty bundle.
func InitRoot(root, trustDomain string) error {
	bundlePath := filepath.Join(root, DirBundleFilename)
	if _, err := os.Stat(bundlePath); err == nil {
		return fmt.Errorf("already initialized: %s", bundlePath)
	}
	if err := os.MkdirAll(filepath.Join(root, servicesDirname), 0o755); err != nil {
		return err
	}
	return writeBundle(bundlePath, Bundle{
		TrustDomain:     trustDomain,
		SpiffileVersion: BundleVersion,
		Identities:      map[string]KeySet{},
	})
}

// AddService generates a keypair for a service and registers its public key
// in the bundle. Idempotent: a service that already has a key keeps it.
func AddService(root, name string) (SpiffeID, error) {
	bundle, err := readBundle(root)
	if err != nil {
		return SpiffeID{}, err
	}
	id, err := ParseSpiffeID(fmt.Sprintf("spiffe://%s/%s", bundle.TrustDomain, name))
	if err != nil {
		return SpiffeID{}, err
	}

	serviceDir := filepath.Join(root, servicesDirname, name)
	keyPath := filepath.Join(serviceDir, DirKeyFilename)
	if _, err := os.Stat(keyPath); err == nil {
		return id, nil
	}

	if err := os.MkdirAll(serviceDir, 0o755); err != nil {
		return SpiffeID{}, err
	}
	pemBytes, err := GeneratePrivateKeyPEM()
	if err != nil {
		return SpiffeID{}, err
	}
	if err := os.WriteFile(keyPath, pemBytes, 0o600); err != nil {
		return SpiffeID{}, err
	}
	if err := os.WriteFile(filepath.Join(serviceDir, DirIDFilename), []byte(id.String()+"\n"), 0o644); err != nil {
		return SpiffeID{}, err
	}

	jwk, err := PublicJWKFromPEM(pemBytes)
	if err != nil {
		return SpiffeID{}, err
	}
	return id, setKeys(root, bundle, id.String(), []JWK{jwk})
}

// RotateService generates a new keypair for a service. With keepOld the
// previous public key stays in the bundle so in-flight tokens and stale
// bundle copies keep working.
func RotateService(root, name string, keepOld bool) (SpiffeID, error) {
	bundle, err := readBundle(root)
	if err != nil {
		return SpiffeID{}, err
	}
	id, err := ParseSpiffeID(fmt.Sprintf("spiffe://%s/%s", bundle.TrustDomain, name))
	if err != nil {
		return SpiffeID{}, err
	}

	keyPath := filepath.Join(root, servicesDirname, name, DirKeyFilename)
	if _, err := os.Stat(keyPath); err != nil {
		return SpiffeID{}, fmt.Errorf("unknown service %q: %w", name, err)
	}

	pemBytes, err := GeneratePrivateKeyPEM()
	if err != nil {
		return SpiffeID{}, err
	}
	newJWK, err := PublicJWKFromPEM(pemBytes)
	if err != nil {
		return SpiffeID{}, err
	}

	keys := []JWK{newJWK}
	if keepOld {
		for _, raw := range bundle.Identities[id.String()].Keys {
			var jwk JWK
			if err := json.Unmarshal(raw, &jwk); err == nil && jwk.Kid != newJWK.Kid {
				keys = append(keys, jwk)
			}
		}
	}

	if err := os.WriteFile(keyPath, pemBytes, 0o600); err != nil {
		return SpiffeID{}, err
	}
	return id, setKeys(root, bundle, id.String(), keys)
}

// RemoveService removes a service's keys from the bundle (revocation) and
// deletes its directory.
func RemoveService(root, name string) error {
	bundle, err := readBundle(root)
	if err != nil {
		return err
	}
	delete(bundle.Identities, fmt.Sprintf("spiffe://%s/%s", bundle.TrustDomain, name))
	if err := writeBundle(filepath.Join(root, DirBundleFilename), bundle); err != nil {
		return err
	}
	return os.RemoveAll(filepath.Join(root, servicesDirname, name))
}

// ServiceEnv returns the environment variables that point a consumer at its
// identity files.
func ServiceEnv(root, name string) map[string]string {
	serviceDir := filepath.Join(root, servicesDirname, name)
	return map[string]string{
		EnvIDFile:     filepath.Join(serviceDir, DirIDFilename),
		EnvKeyFile:    filepath.Join(serviceDir, DirKeyFilename),
		EnvBundleFile: filepath.Join(root, DirBundleFilename),
	}
}

// LoadIdentity loads a provisioned service's Identity directly (tests, tooling).
func LoadIdentity(root, name string) (*Identity, error) {
	env := ServiceEnv(root, name)
	return FromFiles(env[EnvIDFile], env[EnvKeyFile], env[EnvBundleFile])
}

func readBundle(root string) (Bundle, error) {
	return NewFileBundleSource(filepath.Join(root, DirBundleFilename)).Get()
}

func setKeys(root string, bundle Bundle, spiffeID string, jwks []JWK) error {
	keys := make([]json.RawMessage, 0, len(jwks))
	for _, jwk := range jwks {
		raw, err := json.Marshal(jwk)
		if err != nil {
			return err
		}
		keys = append(keys, raw)
	}
	bundle.Identities[spiffeID] = KeySet{Keys: keys}
	return writeBundle(filepath.Join(root, DirBundleFilename), bundle)
}

func writeBundle(path string, bundle Bundle) error {
	doc, err := MarshalBundle(bundle)
	if err != nil {
		return err
	}
	return os.WriteFile(path, doc, 0o644)
}
